### PR TITLE
Refactor/metric

### PR DESCRIPTION
**Pull request by Shylock-Hg**
Make the metric follow the format `<counter_name>.<statistic_type>.<time_range>`, see the comment in `StatsManager` for detail.

### CODE REVIEW

File: src/common/stats/StatsManager.cpp [hunk 1]
```diff
@@ -81,243 +102,216 @@ int32_t StatsManager::registerHisto(folly::StringPiece counterName,
...
+    folly::toLowerAscii(parts[1]);
+    if (parts[1] == "sum") {
+        method = StatsMethod::SUM;
+    } else if (parts[1] == "count") {
+        method = StatsMethod::COUNT;
+    } else if (parts[1] == "avg") {
+        method = StatsMethod::AVG;
+    } else if (parts[1] == "rate") {
+        method = StatsMethod::RATE;
+    } else if (parts[1][0] == 'p') {
```
Comment (laura-ding) [inline src/common/stats/StatsManager.cpp:290]: better to check the size of parts[1][0] firstly.

File: src/common/stats/StatsManager.cpp [hunk 2]
```diff
@@ -81,243 +102,216 @@ int32_t StatsManager::registerHisto(folly::StringPiece counterName,
...
+            case StatsMethod::SUM:
+                // fallthrough
+            case StatsMethod::AVG:
+                // fallthrough
+            case StatsMethod::RATE:
+                // fallthrough
+            case StatsMethod::COUNT:
+                metricValue = readValue(statsName.second, spec.method);
+                break;
+            default:
```
Comment (laura-ding) [inline src/common/stats/StatsManager.cpp:257]: here can delete

File: src/common/stats/StatsManager.cpp (hunk 1 above)
Comment (Shylock-Hg) [reply src/common/stats/StatsManager.cpp]: Fixed.

File: src/common/stats/StatsManager.cpp (hunk 2 above)
Comment (Shylock-Hg) [reply src/common/stats/StatsManager.cpp]: Fixed.

File: src/common/stats/StatsManager.h [hunk 3]
```diff
@@ -114,23 +137,17 @@ class StatsManager final {
     std::unordered_map<std::string, int32_t> nameMap_;
 
     // All time series stats
-    std::vector<
-        std::pair<std::unique_ptr<std::mutex>,
-                  std::unique_ptr<StatsType>
-        >
-    > stats_;
+    std::vector<std::pair<std::unique_ptr<std::mutex>, std::unique_ptr<StatsType>>> stats_;
```
Comment (sherman-the-tank) [inline src/common/stats/StatsManager.h:140]: It is hard to read when all these are on one line

File: src/common/stats/StatsManager.h [hunk 4]
```diff
@@ -114,23 +137,17 @@ class StatsManager final {
...
+    std::vector<MetricSpec> statsSpec_;
 
     // All histogram stats
-    std::vector<
-        std::pair<std::unique_ptr<std::mutex>,
-                  std::unique_ptr<HistogramType>
-        >
-    > histograms_;
+    std::vector<std::pair<std::unique_ptr<std::mutex>, std::unique_ptr<HistogramType>>> histograms_;
```
Comment (sherman-the-tank) [inline src/common/stats/StatsManager.h:144]: ditto

File: src/common/stats/test/StatsManagerTest.cpp [hunk 5]
```diff
@@ -29,27 +30,28 @@ TEST(StatsManager, StatsTest) {
...
+    EXPECT_FALSE(StatsManager::readValue("stat01.Count.3600").ok());
     EXPECT_FALSE(StatsManager::readValue("stat01.Count.4600").ok());
-    EXPECT_EQ(55, StatsManager::readValue("stat01.avg.5").value());
-    EXPECT_EQ(50, StatsManager::readValue("stat01.avg.60").value());
-    EXPECT_EQ(50, StatsManager::readValue("stat01.AVG.600").value());
-    EXPECT_EQ(50, StatsManager::readValue("stat01.Avg.3600").value());
+    EXPECT_FALSE(StatsManager::readValue("stat01.avg.5").ok());
+    EXPECT_FALSE(StatsManager::readValue("stat01.avg.60").ok());
+    EXPECT_FALSE(StatsManager::readValue("stat01.AVG.600").ok());
+    EXPECT_FALSE(StatsManager::readValue("stat01.Avg.3600").ok());
```
Comment (sherman-the-tank) [inline src/common/stats/test/StatsManagerTest.cpp:45]: Why all these become FALSE?

**Review by sherman-the-tank — COMMENTED** [review]
I have a general question about this refactor, what is the goal?

File: src/common/stats/test/StatsManagerTest.cpp (hunk 5 above)
Comment (Shylock-Hg) [reply src/common/stats/test/StatsManagerTest.cpp:45]: They are not registered, different <counter_name>.<statistic_type>.<time_range> label for counter means different counter in this PR.

**Comment by Shylock-Hg** [comment]
Already in #392

**Pull request closed without merging**